### PR TITLE
BHE Pipline Network Coupling through Python Interface

### PR DESCRIPTION
**Pull request by HBShaoUFZ**
In this PR, the feature pipeline network is coupled with the HEAT_TRANSPORT_BHE process through a python interface. This feature is mainly developed by Shuang Chen. 

The benchmark case "3D_3BHE_array" is  located at \Tests\Data\Parabolic\T\3D_3BHE_array
The running of the benchmark requires the following configurations: 

- In the cmake configuration, the option"OGS_USE_PYTHON" is turned on; 
- The execution environment is set to the same folder where the prj file is located; 
- In the python environment, the package Numpy, CoolProp and TESPy have to be installed. 

Source code is ready for review. The documentation of the benchmark case will be uploaded in the next days.

### CODE REVIEW

File: ProcessLib/BoundaryCondition/Python/BHEInflowPythonBoundaryCondition.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        if (dataframe_node_id[i] == boundary_node_id)
+        {
+            bc_values.values[0] = dataframe_Tin_val[i];
+            break;
+        }
+    }
+}
+
+std::unique_ptr<BHEInflowPythonBoundaryCondition>
+createBHEInflowPythonBoundaryCondition(
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/Python/BHEInflowPythonBoundaryCondition.cpp]: Move in own `CreateBHEInflowPythonBoundaryCondition.cpp`. Same with `.h`.

File: ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionModule.cpp [hunk 2]
```diff
@@ -59,5 +59,4 @@ void pythonBindBoundaryCondition(pybind11::module& m)
              &PythonBoundaryConditionPythonSideInterface::getDirichletBCValue);
     pybc.def("getFlux", &PythonBoundaryConditionPythonSideInterface::getFlux);
 }
-
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/Python/PythonBoundaryConditionModule.cpp]: Please, don't change unrelated files.

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp [hunk 3]
```diff
@@ -19,6 +19,8 @@
 #include "HeatTransportBHEProcess.h"
 #include "HeatTransportBHEProcessData.h"
 
+#include <pybind11/pybind11.h>
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: The process should not depend on python, in my opinion. Why is the dependency necessary and can it be moved in the boundary condition?

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp [hunk 4]
```diff
@@ -184,23 +190,82 @@ std::unique_ptr<Process> createHeatTransportBHEProcess(
...
-                                             density_fluid,
-                                             density_gas,
-                                             std::move(bhes)};
+    //! Python object computing BC values.
+    HeatTransportBHEProcessData* process_data;
 
+    if (if_bhe_network_exist_python_bc == false)
+    {
+        process_data =
+            new HeatTransportBHEProcessData(thermal_conductivity_solid,
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: `new` and `delete` are nowadays replaced by `unique_ptr` (or other smart pointers). In the particular case I don't see why a pointer is needed: there is no base class used here.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 5]
```diff
@@ -197,6 +199,44 @@ void HeatTransportBHEProcess::computeSecondaryVariableConcrete(
...
+                std::get<1>(_process_data.py_bc_object->dataframe_network),
+                std::get<2>(_process_data.py_bc_object->dataframe_network));
+        auto const cur_Tin = std::get<2>(tespy_result);
+
+        auto const if_convergence = std::get<1>(tespy_result);
+        if (if_convergence == true)
+            return NumLib::IterationResult::SUCCESS;
+        else
+            for (std::size_t i = 0; i < n_bc_nodes; i++)
+                std::get<1>(_process_data.py_bc_object->dataframe_network)[i] =
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: Braces around (even single line) if/else/for branches/loops missing.

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp [hunk 6]
```diff
@@ -184,23 +190,82 @@ std::unique_ptr<Process> createHeatTransportBHEProcess(
...
+                                            heat_capacity_solid,
+                                            heat_capacity_fluid,
+                                            heat_capacity_gas,
+                                            density_solid,
+                                            density_fluid,
+                                            density_gas,
+                                            std::move(bhes),
+                                            if_bhe_network_exist_python_bc);
+    }
+    // creat a pythonBoundaryCondition object
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: create

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcessData.h [hunk 7]
```diff
@@ -82,6 +89,12 @@ struct HeatTransportBHEProcessData
     std::unordered_map<int, int> _map_materialID_to_BHE_ID;
 
     std::vector<BHE::BHETypes> _vec_BHE_property;
+
+    // if bhe network exists python boundary condition
+    bool if_bhe_network_exist_python_bc;
+
+    //! Python object computing BC values.
+    BHEInflowPythonBoundaryConditionPythonSideInterface* py_bc_object;
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcessData.h]: No plain pointers, please. This is a memory leak.

File: Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+bhe2 = cmp.heat_exchanger_simple('BHE2')
+bhe3 = cmp.heat_exchanger_simple('BHE3')
+
+mg = cmp.merge('merge',num_in = 3)
+
+cons = cmp.heat_exchanger_simple('consumer')
+# two btes subsystems (pipes in parallel flow):
+# 10 pipes for outer ring, 5 pipes inner ring
+#bo = bp('btes outer', 4)
+#bi = bp('btes inner', 2)
```
Comment (endJunction) [inline Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py]: There is quite a bit of commented code in the python modules. Is it needed?

File: Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+# cycle closer (from consumer)
+fc_in = cmp.source('from consumer inflow')
+fc_out = cmp.sink('from consumer outflow')
+
+p1 = cmp.pump('pump')
+
+sp = cmp.splitter('splitter',num_out = 3)
+
+#bhe:
+bhe1 = cmp.heat_exchanger_simple('BHE1')#!name must be same as the user setting in inp_network.py
```
Comment (endJunction) [inline Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py]: Don't assume, the name is same as somewhere else, write an assertion.

**Comment by endJunction** [comment]
Main questions for discussion: what is the difference to the existing PythonBoundaryCondition and why is the BHE network boundary condition not derived from the existing implementation?

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp (hunk 6 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: fixed in commit 87cce675

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp (hunk 5 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: fixed in commit 87cce675

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 10]
```diff
@@ -197,6 +199,44 @@ void HeatTransportBHEProcess::computeSecondaryVariableConcrete(
         getDOFTable(process_id), t, x, _coupled_solutions);
 }
 
+NumLib::IterationResult HeatTransportBHEProcess::postIterationConcreteProcess(
+    GlobalVector const& x)
+{
+    // if the process use python boundary conditon
+    if (_process_data.if_bhe_network_exist_python_bc == false)
+        return NumLib::IterationResult::SUCCESS;
+    else
```
[the remark this answers is not in the document]
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: fixed in commit 87cce675

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 11]
```diff
@@ -9,8 +9,8 @@
 
 #include "HeatTransportBHEProcess.h"
 
-#include <iostream>
 #include <cassert>
+#include <iostream>
```
Comment (TomFischer) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: I don't think the iostream include is needed.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (endJunction) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: It looks like there are more if-else cases with returns.

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp (hunk 6 above)
Comment (endJunction) [reply ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: All typos of this type should be fixed.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 12]
```diff
@@ -239,18 +277,37 @@ void HeatTransportBHEProcess::createBHEBoundaryConditionTopBottom(
...
+                    // Top, inflow.
+                    bcs.addBoundaryCondition(
+                        ProcessLib::createBHEInflowDirichletBoundaryCondition(
+                            get_global_bhe_bc_indices(bc_top_node_id,
+                                                      in_out_component_id),
+                            bhe));
+
+                    // Bottom, outflow.
+                    bcs.addBoundaryCondition(
+                        ProcessLib::createBHEBottomDirichletBoundaryCondition(
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: Code duplication for the bottom BC is not needed.

File: Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py (hunk 8 above)
Comment (ShuangChen88) [reply Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py]: No, it is no needed. The unnecessary commented codes in python modules will be cleaned.

File: Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py (hunk 9 above)
Comment (ShuangChen88) [reply Tests/Data/Parabolic/T/3D_3BHE_array/pre/3bhes.py]: The assertion code will be added to ensure the name of BHE is correct.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp (hunk 12 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: fixed in commit #95f4087el

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp (hunk 11 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: removed.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcessData.h (hunk 7 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcessData.h]: fixed in commit #b23114c6

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp (hunk 4 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: fixed in commit #32114c54

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp (hunk 3 above)
Comment (HBShaoUFZ) [reply ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: This dependency cannot be moved to the boundary condition class. The reason is, the OGS-TESPy coupling is not only executed on the boundary condition level. At the end of each linear solver execution, inside of the HeatTransportBHEProcess::postIterationConcreteProcess function, the Python script will compare the simulated temperatures of each BHE and decide whether the OGS iteration should continue or not. This means, the process itself needs to be linked to pybind11 library.

**Comment by HBShaoUFZ** [comment]
> Main questions for discussion: what is the difference to the existing PythonBoundaryCondition and why is the BHE network boundary condition not derived from the existing implementation?

I just discussed this with Shuang and our answer is "probably not". As you see in my previous answer about "pybind11", the coupling requires actions in the post iteration level. This is beyond the boundary condition class and related directly to the process control. Also, in cases where there are hundreds of BHEs (which we will have), the amount of data exchange is not negligible. In the current version, the data from each BHE are collected altogether in a continuous memory space and exchanged with external Python script. In case we use or derive from the PythonBoundaryCondition class, the exchangeable data will be scattering on each BHE inflow boudary condition and treated one after another. Notice that the determination of whether the iteration should go on or not has to be done only in the postIterationConcrete function. We do not gain anything in our favor if we stick to  PythonBoundaryCondition class or its variant.

File: ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp (hunk 3 above)
Comment (bilke) [reply ProcessLib/HeatTransportBHE/CreateHeatTransportBHEProcess.cpp]: @HBShaoUFZ I have disabled building the process when `OGS_USE_PYTHON=OFF`. This fixed the failing Jenkins jobs.

**Comment by HBShaoUFZ** [comment]
We are not happy with the current status of the code. Dima is correct that the HeatTransportBHE process should not depend on Python interface. Please close this PR for now. Once the code is ready, we will send another PR.
